resources/assert_lib.brs

```diff
@@ -60,3 +61,96 @@ function __formatError(errMessage)
         stack: {}
     }
 end function
+
+sub __deepEquals(actual, expected, error)
+    if __roca_deepEquals(actual, expected) = true then
+        m.__pass()
+    else
+        m.__fail(m.formatError(error))
+    end if
+end sub
+
+'utility function to compare two objects/associative arrays with nested data structures
+function __roca_deepEquals(lhs as object, rhs as object) as boolean
+    ' because Brightsript does not support closures and we are doing recursion,
+    'we want to hold the function parameters in top-level state variables during each execution
+    updatedLeftState = invalid
+    updatedRightState = invalid
+    isEqual = true
+
+    ' base case
+    if type(lhs) <> type(rhs) then return false
+
+    if isArray(lhs) and isArray(rhs) then
+        ' base case
+        if lhs.count() <> rhs.count() then return false
+
+        updatedLeftState = lhs
+        updatedRightState = rhs
+
+        index = 0
+        for each updatedLeft in updatedLeftState
+            updatedRight = updatedRightState[index]
+
+            isEqual = deepEquals(updatedLeft, updatedRight)
+
+            if not isEqual then return false
+
+            index++
+        end for
+    end if
+
+    if isAssociativeArray(lhs) and isAssociativeArray(rhs) then
+        ' base case
+        if lhs.count() <> rhs.count() then return false
+
+        for each key in lhs.keys()
+            ' base case
+            if not rhs.doesExist(key) then return false
+
+            if isArray(lhs[key]) and isArray(rhs[key])then
+                updatedLeftState = lhs[key]
+                updatedRightState = rhs[key]
+
+                isEqual = deepEquals(updatedLeftState, updatedRightState)
+
+                if not isEqual then return false
+            end if
+
+            if isAssociativeArray(lhs[key]) and isAssociativeArray(rhs[key]) then
+                updatedLeftState = lhs[key]
+                updatedRightState = rhs[key]
+
+                isEqual = deepEquals(updatedLeftState, updatedRightState)
+
+                if not isEqual then return false
+            end if
+
+            if isString(lhs[key]) and isString(rhs[key]) then
+                if lhs[key] <> rhs[key] then return false
+            end if
+
+            if isInteger(lhs[key]) and isInteger(rhs[key]) then
+                if lhs[key] <> rhs[key] then return false
+            end if
```

Comment:
Are we missing a few primitive types here? We should be able to compare all numeric types against each other (it doesn't have to be `int` vs `int`!), but I don't see any checks to support things like comparing two functions or comparing to `invalid` primitives.

Consider moving the primitive comparisons (for strings, numbers, `invalid`, raw `function`s, etc.) outside of this `if`.  We might be able to get away with something like:

```brightscript
if type(lhs) <> type(rhs) then return false

if isArray(lhs) and isArray(rhs) then
    ' call __roca_deepEquals on each pair
end if

if isAssociativeArray(lhs) and isAssociativeArrah(rhs) then
    ' compare lhs.count() with rhs.count()
    ' iterate across all keys, calling `__roca_deepEquals(lhs[key], rhs[key])`
end if

' at this point, lhs and rhs are both:
' 1. Strings
' 2. Invalid
' 3. Functions (which always compare against each other as `false`)
' so direct comparisons should be safe

return lhs = rhs

' 4. Some complex type (roDeviceInfo, roDateTime, etc.) that might require special handling (safe to ignore for now)
' 5. Instances of `roSGNode`, which I suspect we can compare using the same logic as associative arrays above

' Print a warning and return `false` I guess, so future upgrades don't suddenly start breaking existing tests
```